File: src/main/java/cube8540/oauth/authentication/credentials/oauth/security/provider/ClientCredentialsAuthenticationProvider.java
package cube8540.oauth.authentication.credentials.oauth.security.provider;

import cube8540.oauth.authentication.credentials.oauth.OAuth2ClientDetails;
import cube8540.oauth.authentication.credentials.oauth.OAuth2ClientDetailsService;
import cube8540.oauth.authentication.credentials.oauth.client.domain.exception.ClientNotFoundException;
import org.springframework.security.authentication.AuthenticationProvider;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.authentication.InternalAuthenticationServiceException;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.crypto.password.PasswordEncoder;

import java.util.Collections;

public class ClientCredentialsAuthenticationProvider implements AuthenticationProvider {

    private final OAuth2ClientDetailsService service;
    private final PasswordEncoder encoder;

    public ClientCredentialsAuthenticationProvider(OAuth2ClientDetailsService service, PasswordEncoder encoder) {
        this.service = service;
        this.encoder = encoder;
    }

    @Override
    public Authentication authenticate(Authentication authentication) throws AuthenticationException {
        try {
            if (authentication.getPrincipal() == null || authentication.getCredentials() == null) {
                throw new BadCredentialsException("principal and credentials is required");
            }

            OAuth2ClientDetails client = service.loadClientDetailsByClientId(authentication.getPrincipal().toString());
            String givenSecret = authentication.getCredentials().toString();
            if (!encoder.matches(givenSecret, client.getClientSecret())) {
                throw new BadCredentialsException("secret does not match stored value");
            }
            return new ClientCredentialsToken(client, client.getClientSecret(), Collections.emptyList());
        } catch (ClientNotFoundException e) {
            throw new BadCredentialsException(e.getMessage());
        } catch (BadCredentialsException | InternalAuthenticationServiceException e) {
            throw e;
        }catch (Exception e) {
            throw new InternalAuthenticationServiceException(e.getMessage(), e);
        }
    }

    @Override
    public boolean supports(Class<?> authentication) {
        return ClientCredentialsToken.class.isAssignableFrom(authentication);
    }
}

File: src/main/kotlin/cube8540/oauth/authentication/credentials/oauth/security/provider/ClientCredentialsAuthenticationProvider.kt
package cube8540.oauth.authentication.credentials.oauth.security.provider

import cube8540.oauth.authentication.credentials.oauth.error.OAuth2ClientRegistrationException
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2ClientDetailsService
import org.springframework.security.authentication.AuthenticationProvider
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.authentication.InternalAuthenticationServiceException
import org.springframework.security.core.Authentication
import org.springframework.security.crypto.password.PasswordEncoder
import java.util.*

class ClientCredentialsAuthenticationProvider(private val service: OAuth2ClientDetailsService, private val encoder: PasswordEncoder): AuthenticationProvider {

    override fun authenticate(authentication: Authentication): Authentication {
        try {
            if (authentication.principal == null || authentication.credentials == null) {
                throw BadCredentialsException("Principal and credentials is required")
            }

            val client = service.loadClientDetailsByClientId(authentication.principal.toString())
            val givenSecret = authentication.credentials.toString()

            if (!encoder.matches(givenSecret, client.clientSecret)) {
                throw BadCredentialsException("Secret does not match stored value")
            }

            return ClientCredentialsToken(client, client.clientSecret, Collections.emptyList())
        } catch (e: OAuth2ClientRegistrationException) {
            throw BadCredentialsException(e.message)
        } catch (e: Exception) {
            when (e) {
                is BadCredentialsException,
                is InternalAuthenticationServiceException -> throw e
                else -> throw InternalAuthenticationServiceException(e.message, e)
            }
        }
    }

    override fun supports(authentication: Class<*>): Boolean =
        ClientCredentialsToken::class.java.isAssignableFrom(authentication)
}